luigi/scheduler.py

```diff
@@ -591,6 +612,7 @@ def get_work(self, worker, host=None, **kwargs):
                               for worker in self._state.get_active_workers())
 
         tasks = list(self._state.get_pending_tasks())
+        self._state.cascade_disabled_state(tasks)
```

Comment:
1. The problem is that workers that are kept-alive never terminate if any tasks become DISABLED. So although currently luigi won't run a task if its dependencies are DISABLED, it also won't terminate, since it'll never try to run a disabled task, so therefore will never fail, so therefore will never be able to apply retry logic. That's why we need to cascade the DISABLED state - to make kept-alive workers terminate.
2. I don't think we'd need to "undisable" anything. cascade_disabled_state goes through all tasks, and for each one it might update the state depending on the states of its dependencies. Or are you saying there's some way that luigi itself can "undisable" tasks? I've seen things go from FAILED -> PENDING, but never from DISABLED -> PENDING...